### ISSUE TITLE
CI emits `torch` UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars

### ISSUE THREAD

**Issue by albertvillanova** [huggingface/trl#7018]
### What

CI emits `UserWarning`s from `torch` in the SFT tests, in the **multi-GPU** slow tests job only:

```
tests/test_sft_trainer.py: 16 warnings
  /__w/trl/trl/.venv/lib/python3.12/site-packages/torch/autograd/function.py:625: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
    return super().apply(*args, **kwargs)  # type: ignore[misc]
```

CI run: https://github.com/huggingface/trl/actions/runs/33618846043/job/100210907376

Affected tests: 16 warnings, all from `tests/test_sft_trainer.py` (pytest collapses them to the file, since the same warning is recorded in more than ten locations).

The single-GPU job emits none of them.

The tests themselves pass; for now this is warning noise only.

### When it started

The warnings appeared in the multi-GPU slow tests job on 2026-08-20. They are absent from every earlier run of that job and present in every run since:

| commit | run | warnings |
|---|---|---|
| `ce01ad43` | https://github.com/huggingface/trl/actions/runs/32359523680 | 0 |
| `4d37e9e2` | https://github.com/huggingface/trl/actions/runs/32394938434 | 16 |

The two runs in between (`3f738b90`, `29b75df4`) died before pytest printed a warnings summary, see #6836.

Nothing changed in the installed dependency versions across that boundary.

### Next steps

Investigate the root cause, then decide whether anything is actionable on the TRL side and/or should be reported upstream.

**Comment by albertvillanova** [comment]
## Root cause

The warning is emitted unconditionally by `torch.nn.parallel._functions.Gather.forward` whenever every input tensor is 0-dim and `dim == 0` ([`v2.13.0`](https://github.com/pytorch/pytorch/blob/v2.13.0/torch/nn/parallel/_functions.py), the exact version the job installs):

```python
class Gather(Function):
    @staticmethod
    def forward(ctx, target_device, dim, *inputs):
        ...
        if all(t.dim() == 0 for t in inputs) and dim == 0:
            inputs = tuple(t.view(1) for t in inputs)
            warnings.warn(
                "Was asked to gather along dimension 0, but all "
                "input tensors were scalars; will instead unsqueeze "
                "and return a vector.",
                stacklevel=2,
            )
            ctx.unsqueezed_scalar = True
```

The reported frame is `torch/autograd/function.py:625` (`Function.apply`) rather than `_functions.py`, because `Gather` is an autograd `Function` and the warning is raised inside its `forward`.

Full call chain:

1. `transformers.Trainer._wrap_model` wraps the model in `nn.DataParallel` ([`trainer.py#L2470`](https://github.com/huggingface/transformers/blob/v5.16.1/src/transformers/trainer.py#L2470) in the installed `v5.16.1`):

   ```python
   elif self.args.n_gpu > 1:
       model = nn.DataParallel(model)
   ```

2. `nn.DataParallel.forward` replicates the module across devices, runs the replicas, then gathers their outputs onto the output device: `self.gather(outputs, self.output_device)`.
3. `torch.nn.parallel.scatter_gather.gather` -> `gather_map` recurses into each replica's output, a `CausalLMOutputWithPast`, and calls `Gather.apply(target_device, dim, *outputs)` per field with `dim=0`.
4. For the `loss` field, every replica contributes a 0-dim scalar, so the `all(t.dim() == 0)` branch is taken and the warning fires.

So the warning is raised entirely inside torch's `DataParallel` implementation. TRL never calls a gather here, and never chooses the `dim`.

## Why it appeared now

Not a dependency change. The installed versions are identical on both sides of the boundary (`torch==2.13.0`, `transformers==5.15.1`, `accelerate==1.14.0`, `peft==0.20.0`).

The cause is #6741 (`1701d3eb`, "Move GPU CI from T4 to L40S", merged 2026-08-20 11:57 UTC), which changed the runner group of the multi-GPU slow tests job:

```diff
   run_all_tests_multi_gpu:
     runs-on:
-      group: aws-g4dn-2xlarge
+      group: aws-g5-12xlarge-cache
```

`g4dn.2xlarge` has a **single** T4, so `CUDA_VISIBLE_DEVICES: "0,1"` resolved to one visible device, `args.n_gpu` was 1, and `nn.DataParallel` was never used. `g5.12xlarge` has 4 A10G, so two devices are genuinely visible and the job became multi-GPU for the first time.

Direct evidence: on `ce01ad43`, the last commit before the move, the `single_gpu` and `multi_gpu` jobs of run [32359523680](https://github.com/huggingface/trl/actions/runs/32359523680) returned byte-identical results, `33 passed, 65 skipped, 2564 deselected`, despite differing in `CUDA_VISIBLE_DEVICES`. Before the move, both jobs ran the same instance type with the same one GPU.

Bisect over the per-commit runs of the multi-GPU job:

| commit | contains #6741 | run | gather warnings | skipped |
|---|---|---|---|---|
| `24db2a82` | no | [31371150907](https://github.com/huggingface/trl/actions/runs/31371150907) | 0 | 72 |
| `06c89977` | no | [32348545023](https://github.com/huggingface/trl/actions/runs/32348545023) | 0 | 65 |
| `ce01ad43` | no | [32359523680](https://github.com/huggingface/trl/actions/runs/32359523680) | 0 | 65 |
| `3f738b90` | yes | [32366421358](https://github.com/huggingface/trl/actions/runs/32366421358) | job died before the summary | n/a |
| `29b75df4` | yes | [32369658762](https://github.com/huggingface/trl/actions/runs/32369658762) | job died before the summary | n/a |
| **`4d37e9e2`** | yes | [32394938434](https://github.com/huggingface/trl/actions/runs/32394938434) | **16** | 42 |
| `47d20ec6` | yes | [32414561886](https://github.com/huggingface/trl/actions/runs/32414561886) | 16 | 42 |
| `e604ec63` | yes | [32497847926](https://github.com/huggingface/trl/actions/runs/32497847926) | 16 | 42 |
| `6630e17a` | yes | [33162166431](https://github.com/huggingface/trl/actions/runs/33162166431) | 16 | 44 |
| `323efb28` | yes | [33618543702](https://github.com/huggingface/trl/actions/runs/33618846043) | 16 | 44 |

The drop in skips (65 -> 42) is the same effect from the other side: the multi-GPU tests that had been silently skipping on a one-GPU runner started to run.

The two runs that died are #6836, the segfault in the same job. Same trigger: that issue is where `xfail_data_parallel` (`tests/testing_utils.py`) comes from, and its comment already records that `Trainer` wraps the model in `nn.DataParallel` as soon as more than one accelerator is visible.

## Is anything actionable on the TRL side?

No, not for the warning itself. What I checked:

- TRL does not call `gather` on the loss anywhere on this path, and does not pass a `dim`. The gather is chosen by `DataParallel`.
- Nothing in TRL can make the per-replica loss non-scalar. The scalar loss is what `DataParallel` is designed to gather, and the resulting vector is what `Trainer` then averages (`if self.args.n_gpu > 1: loss = loss.mean()`, [`trainer.py#L1957`](https://github.com/huggingface/transformers/blob/v5.16.1/src/transformers/trainer.py#L1957)).
- `transformers` already special-cases `DataParallel` next door for `num_items_in_batch` ([`trainer.py#L2203`](https://github.com/huggingface/transformers/blob/v5.16.1/src/transformers/trainer.py#L2203)) without touching the loss, so there is no knob there either.
- No `SFTTrainer` or `SFTConfig` option changes the wrapping. It is driven solely by `args.n_gpu > 1` with no distributed launcher.

The only way to stop the warning at the source is to stop exercising `nn.DataParallel` in that job, which is #6836's subject: TRL trainers do not support it, since they rewire `forward` on a single module instance and `DataParallel` breaks that by replicating the module. That is a CI design decision, out of scope for warning noise, and while #6836 is open the warning is unavoidable.

## Not reproducible locally

`Gather.forward` raises on CPU tensors:

```python
if not all(i.device.type != "cpu" for i in inputs):
    raise AssertionError("Gather function not implemented for CPU tensors")
```

So the warning needs at least two visible CUDA devices and cannot be reproduced on a CPU-only machine.

## Proposed solution

Filter it, and remove the filter once the multi-GPU CI stops exercising `nn.DataParallel` (#6836).

**Comment by albertvillanova** [comment]
Addendum, for accuracy: I abridged the `_wrap_model` snippet above. The wrapping is guarded on quantization, it is not a bare `elif`:

```python
# Multi-gpu training, quantized models do not support DP
if (
    self.args.n_gpu > 1
    and not getattr(model, "is_loaded_in_8bit", False)
    and not getattr(model, "is_loaded_in_4bit", False)
):
    model = nn.DataParallel(model)
```

So the 4-bit SFT slow tests (`test_sft_trainer_transformers_mp_gc_peft_qlora`, `test_sft_trainer_with_chat_format_qlora`) are never wrapped in `nn.DataParallel` and do not contribute to the 16 warnings. Only the non-quantized trainings in `TestSFTTrainerSlow` do. This does not change the conclusion.

**Comment by albertvillanova** [comment]
## Upstream status

**Nothing filed, deliberately.**

The warning belongs to `torch.nn.parallel._functions.Gather`, and it is unchanged there: the code is identical at `v2.13.0` (the version CI installs) and on `pytorch/main`, so the latest release and `main` are both affected, and no upstream fix is pending.

Existing reports, both against `transformers` rather than `torch`, both closed with no code change:

- https://github.com/huggingface/transformers/issues/852 (2019). Answer from a maintainer: "It should be fine. There are probably your output losses."
- https://github.com/huggingface/transformers/issues/14128 (2021). Closed as stale.

No `pytorch` issue exists for it.

I decided against opening one:

- `nn.DataParallel` is soft-deprecated. The torch docs steer everyone to `DistributedDataParallel`, and the warning has been unchanged for years.
- The warning is arguably intentional rather than a bug: it tells the caller that the returned shape differs from what a plain `gather` along dim 0 would give, since the scalars had to be unsqueezed first.
- I could not attach a run-verified reproduction anyway. `Gather.forward` raises on CPU tensors, so reproducing it needs two visible CUDA devices.

Filing would be noise on an API upstream no longer develops. On our side the warning goes away when the multi-GPU CI stops exercising `nn.DataParallel` (#6836), which is what the filter should be tied to.

**Comment by albertvillanova** [comment]
## What we do on the TRL side

Filter it in `pyproject.toml`. There is nothing to fix in TRL: the gather is entirely inside `nn.DataParallel`, and the wrapping is decided by `Trainer` from `n_gpu > 1`.

Tied to #6836 rather than to a dependency version: the warning disappears once the multi-GPU CI stops exercising `nn.DataParallel`.

**Comment by albertvillanova** [comment]
I've opened:
- #7019
